### PR TITLE
Fix thunks in Shelley

### LINKED ISSUE #1558: Run Shelley ledger integration with NoUnexpectedThunks checks enabled
Marking as high priority even for the testnet, because if memory usage just blows up, the testnet won't be very usable.

### PR DESCRIPTION
**Pull request by mrBliss**
Fixes #1558.

### CODE REVIEW

**Comment by mrBliss** [comment]
This depends on https://github.com/input-output-hk/cardano-ledger-specs/pull/1343, but to avoid breaking/stalling things at the end of this sprint, this non-urgent change can wait until next sprint.

**Comment by mrBliss** [comment]
https://github.com/input-output-hk/cardano-ledger-specs/pull/1343 got merged, but some unrelated changes got merged before it, which haven't been propagated to `ouroboros-network` (that's the cost of multiple repos :disappointed:). So this is waiting for #1947.

File: ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 1]
```diff
@@ -15,6 +15,11 @@
 {-# LANGUAGE TypeFamilies               #-}
 {-# LANGUAGE UndecidableInstances       #-}
 
+{-# OPTIONS_GHC -fno-strictness #-}
+-- NOTE: With @-fstrictness@ optimisation (enabled by default for -O1), we get
+-- an unexplained thunk in 'KnownIntersectionState' and thus a space leak. See
+-- #1356.
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:21]: :-(

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
Let's hope CI won't block us too long this time (it did in #1947) :weary:

**Comment by mrBliss** [comment]
bors cancel

**Comment by mrBliss** [comment]
(Bors is running a job for this PR and #1962, which was closed, so there's no point in continuing with that job :weary:)

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
bors ping

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**